backends/guest/common/read.c

```diff
@@ -186,30 +177,31 @@ int print_variables(const uint8_t *buffer, size_t buffer_size, const char *var_n
 				break;
 			case ST_SBAT:
 				printf("\tData: ");
-				print_raw((char *)cert, cert_size);
+				print_raw((char *)curr_esd.raw, esd_data_size);
+				break;
 			case ST_DELETE:
 				printf("\tDELETE-MSG: ");
-				print_raw((char *)cert, cert_size);
+				print_raw((char *)curr_esd.raw, esd_data_size);
+				break;
 			default:
-				prlog(PR_ERR, "ERROR: invalid signature type\n");
-				goto outside;
+				prlog(PR_ERR, "ERROR: invalid signature type = %d\n", sig_type);
+				break;
 			}
 
-			count++;
-			esl_size -= signature_size;
-			sig_offset += signature_size;
+			rc = next_esd_from_esl(curr_esl.raw, &curr_esd.raw, &esd_data_size,
+					       &esd_owner);
+			if (rc) {
+				prlog(PR_ERR, "Error reading next esd (%zu), rc = %d\n", count, rc);
+			}
+		}
+		rc = next_esl_from_buffer(buffer, buffer_size, &curr_esl.raw, &esl_data_size);
+		if (rc) {
+			prlog(PR_ERR, "Error reading next esl (%zu) from buffer: %d\n", count, rc);
+			return rc;
 		}
-	outside:
-
-		/* we read all esl_size bytes so iterate to next esl */
-		esl_data += next_esl_size;
-		esl_data_size -= next_esl_size;
 	}
 
 	printf("\tFound %zu ESL's\n\n", count);
 
-	if (x509)
-		crypto_x509_free(x509);
```

Comment:
Removing this exposes a possible leak if `print_cert_info` returns nonzero. Consider just adding a free call in the error block before the break.

Reply:
Fixed by always deallocating immediately after its only use. Related, return from the function if printing the cert fails. I don't really like the idea of limping along if anything fails to print.